### ISSUE TITLE
LISA: layerwise importance sampling (random per-step layer activation) — follow-up to native Spectrum

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#267]
## Context

Follow-up to #266 (native Spectrum). LISA — Layerwise Importance Sampled AdamW (arXiv:2403.17919) — is the dynamic cousin of Spectrum's static layer selection: instead of picking high-SNR layers once before training, LISA randomly activates a small set of decoder layers every N steps (embeddings + head always trainable), giving full-FT-like quality at LoRA-like memory.

Kept OUT of the #266 release deliberately: LISA lives inside the training loop (per-step freeze/unfreeze callback + optimizer-state handling), which is a different review surface from Spectrum's offline weight scan. The plan's parity table lists LISA as a v0.74-adjacent vignette.

## Proposed fix path

1. `TrainingConfig.lisa_enabled: bool` + `lisa_num_layers: int [1, 64]` + `lisa_interval_steps: int [1, 1e6]` (bool-as-int rejection per project policy). Cross-validators: requires SFT-family task, mutually exclusive with `lora`, `unfrozen_parameters` (#266), `freeze_layers`, `relora_steps`.
2. `utils/lisa.py` — `LisaCallback` (HF TrainerCallback): on every `lisa_interval_steps`, re-sample `lisa_num_layers` decoder layers uniformly, freeze the rest; embeddings + lm_head stay trainable; optimizer state for re-frozen layers cleared (mirrors the v0.39 ReLoRA reset policy). Arch resolution reuses the `_locate_decoder_layers` family (Llama + GPT-2 Conv1D aware).
3. `attach_lisa_callback(trainer, tcfg)` helper in `utils/peft_wiring.py` following the attach-helper pattern; wired into the SFT-family trainers.

## Acceptance criteria

- [ ] Schema fields + cross-validators with named error messages
- [ ] Callback re-samples on the configured cadence; frozen/unfrozen counts verified per step in unit tests
- [ ] Optimizer state cleared for re-frozen layers (no stale Adam moments)
- [ ] Live smoke on SmolLM2-135M (RTX 3050): trains end-to-end, loss decreases, peak VRAM below full-FT baseline
- [ ] Mutual-exclusion matrix tested (lora / unfrozen_parameters / freeze_layers / relora)

**Comment by MakazhanAlpamys** [comment]
Shipped in v0.71.34. See [release notes](https://github.com/MakazhanAlpamys/Soup/releases/tag/v0.71.34).

LISA (Layerwise Importance Sampled AdamW, arXiv:2403.17919) is live: enable with `training.lisa_enabled: true` (`lisa_num_layers`, `lisa_interval_steps`) on an `sft` + transformers + text + `quantization: none` run. `LisaCallback` re-samples a random decoder-layer set every N steps (embeddings + head always trainable) and clears optimizer state for re-frozen layers; `_validate_lisa_compat` gates it and enforces mutual exclusion with LoRA / freeze / Spectrum.

All 5 acceptance criteria met (schema + named errors, per-step re-sample + counts, opt-state reset, mutual-exclusion matrix, live SmolLM2-135M train). VRAM-win magnitude at 7B+ is a documented scale follow-up (filed separately).

Try:
```yaml
task: sft
training:
  quantization: none
  lisa_enabled: true
  lisa_num_layers: 2
  lisa_interval_steps: 20
```
